### ISSUE TITLE
neural_doodle.py ignores some region when using certain style images

### ISSUE THREAD

**Issue by titu1994** [fchollet/keras#3731]
Seems that some styles don't work well with the script for some reason. I utilized the "Starry Night" by Van Gogh as the style image, with the mask provided at https://github.com/DmitryUlyanov/online-neural-doodle , along with the target mask as shown below.

For some reason, the blue swirls (red colored region) are completely ignored. The following image is after 80 iterations.
![neural doodle](https://cloud.githubusercontent.com/assets/3048602/18378687/05dd4600-768c-11e6-8c67-820e275b9192.png)

Changing the color of the masks, or slightly altering the mask itself has no noticeable effect on the final result. The following image is after 10 iterations.
[Neural Doodle with different mask color](http://imgur.com/a/0CYLB)

As can be seen in the 80 iterations image, the buildings and the dark brown tower, as well as the swirls in the sky are drawn correctly, but the blue swatches (masked in red) are just drawn in a gray color. This gray color does not change in any of the 80 iterations.

**Comment by dolaameng** [comment]
Hi @titu1994  thanks for reporting this. Yes I got similar results to yours. I suspect it is because the red and green region is kinda overlapped with other regions. And as a consequence, applying a series of maxpooling on the mask may not work as expected - the regions will be squeezed together? 

I am not really sure. What do you think? Have you tried any target mask without much overlapping? Or have you tried the samples on the original Torch code?

**Comment by titu1994** [comment]
@dolaameng I tried a few things, such as increasing the ncolors to 5 and 6. With 6, it started to produce the blue patches but only in the first 5-7 epochs, then no further updates to that region in next 75 epochs.

I think you are applying average pooling to the masks instead of max pooling? This does decrease quality in the original neural style scripts, even though overall loss value is far less. It is true that the red and green region is overlapping to some degree.

I did try one test with 20 iterations with only 3 colors (I changed the red to green color). It did draw a few small blue swirls in a random manner. I guess the region simply does not contain enough information after pooling operations.

I am working on Windows so Torch and Lua are not available to me. However the online neural doodle link I posted above does use the mask with 4 colors and is able to partially recreate the entire "Starry Night" image (although the sky has no swirl structure)

**Comment by dolaameng** [comment]
@titu1994 : that's interesting. Yes I tried the [online demo](http://likemo.net/) and it seems to work better. Actually Dmitry did mention in [his blog](http://dmitryulyanov.github.io/feed-forward-neural-doodle/) to use a 3x3 mean filter for mask to match convolution and an average pooling for image pooling. But I chose to use averagingpool to match both convolution and pooling for simplicity and it seemed to generate similar results for Monet and Renoir. Do you think it matters in this example?

And another consideration is that I used the area of the region to weigh different parts in the style loss, as suggested in the blog. And you can still that actually made the sky quickly dominate and other parts hard to update. If I turn off the weight, it seems to help a little bit. e..g, this is what I got after 60 iterations.

![starry_at_iteration_60](https://cloud.githubusercontent.com/assets/36718/18392218/9270dc10-76e3-11e6-960b-86637cf6a06f.png)

You can see it is still blur, but all parts are "equally" blur. I will try to look into it in the following couple of days.

**Comment by dolaameng** [comment]
I got better results by turning off the "region area based weight" in style loss, and using a bigger tv_loss_weight (=0.01). Now I tend to suspect more that the loss weight to be the culprit. See results after 60 iterations now.

![starry_at_iteration_60](https://cloud.githubusercontent.com/assets/36718/18393198/167fa81c-76e7-11e6-9327-e6d0f887fbe4.png)

**Comment by titu1994** [comment]
@dolaameng Very nice results ! Not only did the swirl pattern become more prominent, but it also properly styled the red masked region. 

Is the region area loss important according to the paper ? Visyally it's a striking difference.

**Comment by dolaameng** [comment]
@titu1994 : using the ratio of region area (wrt the total image size) is [recommended](http://dmitryulyanov.github.io/feed-forward-neural-doodle/). However, the [current implementation](https://github.com/fchollet/keras/blob/master/examples/neural_doodle.py#L237) is only an approximation of the ratio because the ratio is always with respect to [original image size](https://github.com/fchollet/keras/blob/master/examples/neural_doodle.py#L237). I don't know what effect this will result in, but it is really interesting to observe test cases to cause such a difference. 

I will find some time to experiment a little more before making a fix today, to see how this style loss influence the generation in practice. Currently the sky generation still looks more appealing than other parts in terms of details (e.g., colors?) - I don't know if it's because of its size or something else. It is an interesting test case - as you mentioned the "swirl pattern" is also gone in the generation by the [online demo](http://likemo.net/). 

All suggestions/explorations are welcomed.

**Comment by titu1994** [comment]
@dolaameng I don't think it will affect the outcome much, but in the original blog [Neural doodles](http://dmitryulyanov.github.io/feed-forward-neural-doodle/), the Cl,r loss is normalized by dividing it with the num_rows and num_cols. Perhaps this normalization is more important that what is immediately noticeable? 

I say this because while developing a fast neural style model (similar to keras-rtst, but for Keras 1.0+), I made the mistake of not normalizing the loss values correctly and it caused the validation images to completely lose detail (or it may be some other problem, I still can't get it to work properly). Similarly, I think we should test by normalizing the sum component of the loss. It may play an important role.

**Comment by dolaameng** [comment]
Hi @titu1994 : I have figured out a style loss definition that is more consistent with different style layers. The new loss seems to work with all the tested images so far. Its convergence is also fast - for your starry image example it can generate good quality results in 5 iterations. Personally I also found the new results more appealing because it reveals more details. See the results of three images of 50 iterations.

![monet__at_iteration_49](https://cloud.githubusercontent.com/assets/36718/18410914/fab2e438-779e-11e6-933b-39ba317aec86.png)

![renoir__at_iteration_49](https://cloud.githubusercontent.com/assets/36718/18410917/08d3ea8a-779f-11e6-90f0-e1346d81d042.png)

![starry_at_iteration_49](https://cloud.githubusercontent.com/assets/36718/18410918/0c94980e-779f-11e6-9ce4-ed7b67fcb3b9.png)

I will create a PR to update it. Here are what I have learned so far,

1. Normalizing different features from different style layers is probably more important than normalizing regions of different sizes.
2. For best qualities, the normalization factors for each layer should involve (1) layer size w x h (2) # of layer channels.
3. Different layers of CNN do capture different types of patterns, e.g., different patterns involving different sizes - taking starry image as an example, the big circle stars in the sky v.s. the small buildings of square. If those features are not normalized properly, the generated doodle might be dominated by certain features like what you have observed.

Thank you @titu1994  @bmaltais for coming with the test example!
